### ISSUE TITLE
Testing mocked method invocations with withConsecutive and callback will call callback twice

### ISSUE THREAD

**Issue by marc-mabe** [sebastianbergmann/phpunit#4930]
| Q                   | A
| --------------------| ---------------
| PHPUnit version     | 9.5.18
| PHP version         | 8.1.3
| Installation Method | Composer

#### Summary

Testing mocked method invocations with `withConsecutive` and `callback` will call callback twice.
My my real world case I'm first expecting an entity with specific state to be persistent and secondly another entity to be persisted
but the callback gets called twice, first with my expected entity object and later on with an emptied entity object.
I wasn't able to reproduce the second emptied entity case but here it's the duplicated  call case.

#### Current behavior

The callback provided gets called twice. First on expected method invocation and later on (I think after the test) again.

```
./vendor/bin/phpunit 'tests/MyTest.php'
PHPUnit 9.5.18 #StandWithUkraine

Testing App\Tests\MyTest
Fobject(App\Tests\FirstArg)#91 (1) {
  ["myProp"]=>
  object(stdClass)#70 (1) {
    ["test"]=>
    string(5) "first"
  }
}
Exception in /app/tests/MyTest.php:42
Stack trace:
#0 /app/vendor/phpunit/phpunit/src/Framework/Constraint/Callback.php(50): App\Tests\MyTest::App\Tests\{closure}(Object(App\Tests\FirstArg))
#1 /app/vendor/phpunit/phpunit/src/Framework/Constraint/Constraint.php(46): PHPUnit\Framework\Constraint\Callback->matches(Object(App\Tests\FirstArg))
#2 /app/vendor/phpunit/phpunit/src/Framework/MockObject/Rule/ConsecutiveParameters.php(125): PHPUnit\Framework\Constraint\Constraint->evaluate(Object(App\Tests\FirstArg), '...')
#3 /app/vendor/phpunit/phpunit/src/Framework/MockObject/Rule/ConsecutiveParameters.php(77): PHPUnit\Framework\MockObject\Rule\ConsecutiveParameters->verifyInvocation(Object(PHPUnit\Framework\MockObject\Invocation), 0)
#4 /app/vendor/phpunit/phpunit/src/Framework/MockObject/Matcher.php(137): PHPUnit\Framework\MockObject\Rule\ConsecutiveParameters->apply(Object(PHPUnit\Framework\MockObject\Invocation))
#5 /app/vendor/phpunit/phpunit/src/Framework/MockObject/InvocationHandler.php(122): PHPUnit\Framework\MockObject\Matcher->invoked(Object(PHPUnit\Framework\MockObject\Invocation))
#6 /app/vendor/phpunit/phpunit/src/Framework/MockObject/MockClass.php(51) : eval()'d code(24): PHPUnit\Framework\MockObject\InvocationHandler->invoke(Object(PHPUnit\Framework\MockObject\Invocation))
#7 /app/tests/MyTest.php(74): Mock_Test_793f39d5->myfunc(Object(App\Tests\FirstArg))
#8 /app/vendor/phpunit/phpunit/src/Framework/TestCase.php(1545): App\Tests\MyTest->testWithConsecutiveAndCallback()
#9 /app/vendor/phpunit/phpunit/src/Framework/TestCase.php(1151): PHPUnit\Framework\TestCase->runTest()
#10 /app/vendor/phpunit/phpunit/src/Framework/TestResult.php(726): PHPUnit\Framework\TestCase->runBare()
#11 /app/vendor/phpunit/phpunit/src/Framework/TestCase.php(903): PHPUnit\Framework\TestResult->run(Object(App\Tests\MyTest))
#12 /app/vendor/phpunit/phpunit/src/Framework/TestSuite.php(677): PHPUnit\Framework\TestCase->run(Object(PHPUnit\Framework\TestResult))
#13 /app/vendor/phpunit/phpunit/src/TextUI/TestRunner.php(673): PHPUnit\Framework\TestSuite->run(Object(PHPUnit\Framework\TestResult))
#14 /app/vendor/phpunit/phpunit/src/TextUI/Command.php(143): PHPUnit\TextUI\TestRunner->run(Object(PHPUnit\Framework\TestSuite), Array, Array, true)
#15 /app/vendor/phpunit/phpunit/src/TextUI/Command.php(96): PHPUnit\TextUI\Command->run(Array, true)
#16 /app/vendor/phpunit/phpunit/phpunit(98): PHPUnit\TextUI\Command::main()
#17 /app/vendor/bin/phpunit(120): include('...')
#18 {main}object(App\Tests\SecondArg)#92 (1) {
  ["myProp"]=>
  object(stdClass)#89 (1) {
    ["test"]=>
    string(6) "second"
  }
}
Exception in /app/tests/MyTest.php:53
Stack trace:
#0 /app/vendor/phpunit/phpunit/src/Framework/Constraint/Callback.php(50): App\Tests\MyTest::App\Tests\{closure}(Object(App\Tests\SecondArg))
#1 /app/vendor/phpunit/phpunit/src/Framework/Constraint/Constraint.php(46): PHPUnit\Framework\Constraint\Callback->matches(Object(App\Tests\SecondArg))
#2 /app/vendor/phpunit/phpunit/src/Framework/MockObject/Rule/ConsecutiveParameters.php(125): PHPUnit\Framework\Constraint\Constraint->evaluate(Object(App\Tests\SecondArg), '...')
#3 /app/vendor/phpunit/phpunit/src/Framework/MockObject/Rule/ConsecutiveParameters.php(77): PHPUnit\Framework\MockObject\Rule\ConsecutiveParameters->verifyInvocation(Object(PHPUnit\Framework\MockObject\Invocation), 1)
#4 /app/vendor/phpunit/phpunit/src/Framework/MockObject/Matcher.php(137): PHPUnit\Framework\MockObject\Rule\ConsecutiveParameters->apply(Object(PHPUnit\Framework\MockObject\Invocation))
#5 /app/vendor/phpunit/phpunit/src/Framework/MockObject/InvocationHandler.php(122): PHPUnit\Framework\MockObject\Matcher->invoked(Object(PHPUnit\Framework\MockObject\Invocation))
#6 /app/vendor/phpunit/phpunit/src/Framework/MockObject/MockClass.php(51) : eval()'d code(24): PHPUnit\Framework\MockObject\InvocationHandler->invoke(Object(PHPUnit\Framework\MockObject\Invocation))
#7 /app/tests/MyTest.php(75): Mock_Test_793f39d5->myfunc(Object(App\Tests\SecondArg))
#8 /app/vendor/phpunit/phpunit/src/Framework/TestCase.php(1545): App\Tests\MyTest->testWithConsecutiveAndCallback()
#9 /app/vendor/phpunit/phpunit/src/Framework/TestCase.php(1151): PHPUnit\Framework\TestCase->runTest()
#10 /app/vendor/phpunit/phpunit/src/Framework/TestResult.php(726): PHPUnit\Framework\TestCase->runBare()
#11 /app/vendor/phpunit/phpunit/src/Framework/TestCase.php(903): PHPUnit\Framework\TestResult->run(Object(App\Tests\MyTest))
#12 /app/vendor/phpunit/phpunit/src/Framework/TestSuite.php(677): PHPUnit\Framework\TestCase->run(Object(PHPUnit\Framework\TestResult))
#13 /app/vendor/phpunit/phpunit/src/TextUI/TestRunner.php(673): PHPUnit\Framework\TestSuite->run(Object(PHPUnit\Framework\TestResult))
#14 /app/vendor/phpunit/phpunit/src/TextUI/Command.php(143): PHPUnit\TextUI\TestRunner->run(Object(PHPUnit\Framework\TestSuite), Array, Array, true)
#15 /app/vendor/phpunit/phpunit/src/TextUI/Command.php(96): PHPUnit\TextUI\Command->run(Array, true)
#16 /app/vendor/phpunit/phpunit/phpunit(98): PHPUnit\TextUI\Command::main()
#17 /app/vendor/bin/phpunit(120): include('...')
#18 {main}object(App\Tests\FirstArg)#91 (1) {
  ["myProp"]=>
  object(stdClass)#70 (1) {
    ["test"]=>
    string(5) "first"
  }
}
Exception in /app/tests/MyTest.php:42
Stack trace:
#0 /app/vendor/phpunit/phpunit/src/Framework/Constraint/Callback.php(50): App\Tests\MyTest::App\Tests\{closure}(Object(App\Tests\FirstArg))
#1 /app/vendor/phpunit/phpunit/src/Framework/Constraint/Constraint.php(46): PHPUnit\Framework\Constraint\Callback->matches(Object(App\Tests\FirstArg))
#2 /app/vendor/phpunit/phpunit/src/Framework/MockObject/Rule/ConsecutiveParameters.php(125): PHPUnit\Framework\Constraint\Constraint->evaluate(Object(App\Tests\FirstArg), '...')
#3 /app/vendor/phpunit/phpunit/src/Framework/MockObject/Rule/ConsecutiveParameters.php(87): PHPUnit\Framework\MockObject\Rule\ConsecutiveParameters->verifyInvocation(Object(PHPUnit\Framework\MockObject\Invocation), 0)
#4 /app/vendor/phpunit/phpunit/src/Framework/MockObject/Matcher.php(232): PHPUnit\Framework\MockObject\Rule\ConsecutiveParameters->verify()
#5 /app/vendor/phpunit/phpunit/src/Framework/MockObject/InvocationHandler.php(174): PHPUnit\Framework\MockObject\Matcher->verify()
#6 /app/vendor/phpunit/phpunit/src/Framework/MockObject/Api/Api.php(86): PHPUnit\Framework\MockObject\InvocationHandler->verify()
#7 /app/vendor/phpunit/phpunit/src/Framework/TestCase.php(2002): Mock_Test_793f39d5->__phpunit_verify(true)
#8 /app/vendor/phpunit/phpunit/src/Framework/TestCase.php(1152): PHPUnit\Framework\TestCase->verifyMockObjects()
#9 /app/vendor/phpunit/phpunit/src/Framework/TestResult.php(726): PHPUnit\Framework\TestCase->runBare()
#10 /app/vendor/phpunit/phpunit/src/Framework/TestCase.php(903): PHPUnit\Framework\TestResult->run(Object(App\Tests\MyTest))
#11 /app/vendor/phpunit/phpunit/src/Framework/TestSuite.php(677): PHPUnit\Framework\TestCase->run(Object(PHPUnit\Framework\TestResult))
#12 /app/vendor/phpunit/phpunit/src/TextUI/TestRunner.php(673): PHPUnit\Framework\TestSuite->run(Object(PHPUnit\Framework\TestResult))
#13 /app/vendor/phpunit/phpunit/src/TextUI/Command.php(143): PHPUnit\TextUI\TestRunner->run(Object(PHPUnit\Framework\TestSuite), Array, Array, true)
#14 /app/vendor/phpunit/phpunit/src/TextUI/Command.php(96): PHPUnit\TextUI\Command->run(Array, true)
#15 /app/vendor/phpunit/phpunit/phpunit(98): PHPUnit\TextUI\Command::main()
#16 /app/vendor/bin/phpunit(120): include('...')
#17 {main}....                                                               5 / 5 (100%)

Time: 00:00.041, Memory: 12.00 MB

There was 1 failure:

1) App\Tests\MyTest::testWithConsecutiveAndCallback
Expectation failed for method name is "myfunc" when invoked 2 time(s).
$testCallback1 invoked more than once
Failed asserting that 2 is identical to 1.

FAILURES!
Tests: 5, Assertions: 31, Failures: 1.
ERROR: 1
```

#### How to reproduce

```php
class Test {
    public function myfunc(...$args) {}
}
class FirstArg {}
class SecondArg {}

class MyHandlerTest extends TestCase
{
    public function testWithConsecutiveAndCallback()
    {
        $testCallback1Runs = 0;
        $testCallback1 = static function (FirstArg $arg) use (&$testCallback1Runs) {
            $testCallback1Runs++;

            var_dump($arg);
            echo new \Exception();
            self::assertSame(1, $testCallback1Runs, '$testCallback1 invoked more than once');

            return true;
        };

        $testCallback2Runs = 0;
        $testCallback2 = static function (SecondArg $arg) use (&$testCallback2Runs) {
            $testCallback2Runs++;

            var_dump($arg);
            echo new \Exception();
            self::assertSame(1, $testCallback2Runs, '$testCallback2 invoked more than once');

            return true;
        };

        $mock = $this->createMock(Test::class);
        $mock
            ->expects(static::exactly(2))
            ->method('myfunc')
            ->withConsecutive(
                [self::callback($testCallback1)],
                [self::callback($testCallback2)]
            );

        $firstArg = new FirstArg();
        $firstArg->myProp = (object)['test' => 'first'];

        $secondArg = new SecondArg();
        $secondArg->myProp = (object)['test' => 'second'];

        $mock->myfunc($firstArg);
        $mock->myfunc($secondArg);
    }
}
```

#### Expected behavior

Each provided callback should only be called once and the test should pass.

**Comment by sebastianbergmann** [comment]
It is because of problems such as this that `withConsecutive()` has been deprecated in PHPUnit 10 and is scheduled for removal in PHPUnit 11. See https://github.com/sebastianbergmann/phpunit/issues/4255#issuecomment-636422439 for more details.

TL;DR: Won't fix, because the functionality will be removed anyway.

**Comment by marcospassos** [comment]
@sebastianbergmann, for order sensitive code, like verifying that must traverser call the visitor in the correct order, what are the alternatives if the `onConsecutive` gets deprecated?
